blockingconcurrentqueue.h

```diff
@@ -207,14 +233,22 @@ namespace details
 		        {
 		            oldCount = m_count.load(std::memory_order_relaxed);
 		            if ((oldCount > 0) && m_count.compare_exchange_strong(oldCount, oldCount - 1, std::memory_order_acquire, std::memory_order_relaxed))
-		                return;
+		                return true;
 		            std::atomic_signal_fence(std::memory_order_acquire);     // Prevent the compiler from collapsing the loop.
 		        }
 		        oldCount = m_count.fetch_sub(1, std::memory_order_acquire);
 		        if (oldCount <= 0)
 		        {
-		            m_sema.wait();
+		            if (timeout_usecs >= 0)
+		            {
+		                return m_sema.wait_for_usecs(timeout_usecs);
```

Comment:
Ah, this isn't correct. If the wait on the semaphore times out, `m_count` is still left decremented (indicating that someone is waiting for an element). This essentially imbalances the count of the semaphore after each timeout.